### PR TITLE
e2e tests and examples for inflate helm chart

### PR DESCRIPTION
**Pull request by natasha41575**
This PR adds tests/examples for the following use cases:

- includeCRDs
- remote valuesFile
- valuesInline

It also fixes a bug I found with includeCRDs.

### CODE REVIEW

File: examples/inflate-helm-chart-crds/.expected/config.yaml
Comment (droot) [inline examples/inflate-helm-chart-crds/.expected/config.yaml:17]: /cc @phanimarupaka

File: examples/inflate-helm-chart-crds/README.md
Comment (droot) [inline examples/inflate-helm-chart-crds/README.md:5]: s/demonstrated/demonstrates

File: examples/inflate-helm-chart-crds/README.md
Comment (droot) [inline examples/inflate-helm-chart-crds/README.md:6]: s/while including/that contains

File: examples/inflate-helm-chart-remote-values-file/.expected/exec.sh
Comment (droot) [inline examples/inflate-helm-chart-remote-values-file/.expected/exec.sh:9]: question: values.yaml can exist locally as well ?

File: examples/inflate-helm-chart-remote-values-file/README.md
Comment (droot) [inline examples/inflate-helm-chart-remote-values-file/README.md:5]: s/demonstrated/demonstrates

File: examples/inflate-helm-chart-remote-values-file/README.md
Comment (droot) [inline examples/inflate-helm-chart-remote-values-file/README.md:6]: chart term seems unnecessary here.

**Review by droot — COMMENTED** [review]
minor comments.

File: examples/inflate-helm-chart-remote-values-file/.expected/exec.sh
Comment (natasha41575) [reply examples/inflate-helm-chart-remote-values-file/.expected/exec.sh:9]: yes. The example `inflate-helm-chart-local` covers this

**Comment by natasha41575** [comment]
Regarding the CheckSite failure, @etefera said it doesn't seem like this PR caused it and has suggested we merge after approval even if it is still failing. He will chat about it with @mengqiy when he comes back.

File: examples/inflate-helm-chart-crds/README.md
Comment (droot) [inline examples/inflate-helm-chart-crds/README.md:6]: Question: May be I am too late to comment on this aspect. Did we consider using names other than `inflate` ? I think `render-helm-chart` would have been more intuitive for the end users.

File: examples/inflate-helm-chart-crds/README.md
Comment (droot) [inline examples/inflate-helm-chart-crds/README.md:20]: `You should see rendered chart resources in your local filesystem` ?

File: functions/go/inflate-helm-chart/README.md
Comment (droot) [inline functions/go/inflate-helm-chart/README.md:150]: This is inconsistent with other functions. I would suggest focusing on just the functionConfig here and usage with kpt like we do for other functions. Include an example to demonstrate how to use this function with kustomize ? We can follow the same model for other functions as well. So when we introduce changes in kustomize CLI, those will impact just the kustomize examples so keeping the function pages decoupled from kustomize's changes as well.

**Review by droot — COMMENTED** [review]
Looking good, minor comments.

Also curious, how did you resolve the CI failure related to permission error ?

File: examples/inflate-helm-chart-crds/README.md
Comment (natasha41575) [reply examples/inflate-helm-chart-crds/README.md:6]: I chose `inflate-helm-chart` because that's what the other function was called and the kustomize builtin plugin is also called HelmChartInflationGenerator.

But also considering that the [description for `helm template`](https://helm.sh/docs/helm/helm_template/)  says "locally renders templates", I agree that `render-helm-chart` is more intuitive.

I will change it to `render-helm-chart` and confirm with @mengqiy before release.

File: examples/inflate-helm-chart-crds/README.md
Comment (natasha41575) [reply examples/inflate-helm-chart-crds/README.md]: Done

File: functions/go/inflate-helm-chart/README.md
Comment (natasha41575) [reply functions/go/inflate-helm-chart/README.md]: I have moved the kustomize example to its own example page, decoupling it from the function page.

I can also remove the kustomize example entirely if it is not desired. The main reason for including it was to test `valuesInline`, as there is no way to test this with kpt CLI.

File: examples/render-helm-chart-crds/.expected/config.yaml
Comment (droot) [inline examples/render-helm-chart-crds/.expected/config.yaml:17]: @natasha41575 Can you pl. file a bug for it if you haven't already. Thank you so much!

File: examples/inflate-helm-chart-crds/README.md
Comment (droot) [reply examples/inflate-helm-chart-crds/README.md]: Great. I saw it has been renamed to `render-helm-chart`.

File: functions/go/inflate-helm-chart/README.md
Comment (droot) [reply functions/go/inflate-helm-chart/README.md]: I am happy with the new changes.

**Review by droot — COMMENTED** [review]
lgtm

File: examples/render-helm-chart-crds/.expected/config.yaml
Comment (natasha41575) [reply examples/render-helm-chart-crds/.expected/config.yaml:17]: filed https://github.com/GoogleContainerTools/kpt/issues/2462

Comment (phanimarupaka) [reply examples/render-helm-chart-crds/.expected/config.yaml:17]: @natasha41575 I added the test and it works fine. https://github.com/kubernetes-sigs/kustomize/pull/4134/files

Which version of kpt are you using ? `kpt version` on your local should have version beta2 or above.

Where can I find render-helm-chart function ? I don't see it here https://pantheon.corp.google.com/gcr/images/kpt-fn/GLOBAL

Comment (mengqiy) [reply examples/render-helm-chart-crds/.expected/config.yaml:17]: >Where can I find render-helm-chart function ?

It is being renamed in this PR. You can find it currently as `gcr.io/kpt-fn/inflate-helm-chart:unstable`

File: examples/render-helm-chart-crds/.expected/config.yaml
Comment (mengqiy) [inline examples/render-helm-chart-crds/.expected/config.yaml:19]: I guess we don't have to skip it, right?
Even there may be an issue with the fields ordering, we can tweak the fields ordering in the input to reduce the unnecessary diff.

File: examples/render-helm-chart-crds/.expected/exec.sh
Comment (mengqiy) [inline examples/render-helm-chart-crds/.expected/exec.sh:3]: We should use `--image-pull-policy never` since we always build the images before testing.

File: examples/render-helm-chart-kustomize-inline-values/README.md
Comment (mengqiy) [inline examples/render-helm-chart-kustomize-inline-values/README.md:51]: Can you please talk a bit more about what does it mean when I see this resource in my output?
How can the user confirm if the inline values file is taking effects? What is the differences when using the default values file vs the inline values file?

You have the following in another example. I think it's good enough if you have something similar
```
You should be able to find `replicas: 5` in
file `deployment_test-helloworld-chart.yaml`, which demonstrates that
the correct values file provided by --valuesFile was used.
```

File: examples/render-helm-chart-kustomize-inline-values/kustomization-dir/kustomization.yaml
Comment (mengqiy) [inline examples/render-helm-chart-kustomize-inline-values/kustomization-dir/kustomization.yaml:1]: Do we need to keep `kustomization.yaml` in a separate directory? Can we keep in flat?

File: examples/render-helm-chart-remote-values-file/.expected/exec.sh
Comment (mengqiy) [inline examples/render-helm-chart-remote-values-file/.expected/exec.sh:3]: --image-pull-policy never

File: examples/render-helm-chart-remote-values-file/README.md
Comment (mengqiy) [inline examples/render-helm-chart-remote-values-file/README.md:24]: Remove `--image-pull-policy ifNotPresent` here.

File: examples/render-helm-chart-remote-values-file/README.md
Comment (mengqiy) [inline examples/render-helm-chart-remote-values-file/README.md:17]: You don't have to do this if you run `kpt fn eval render-helm-chart-remote-values-file`.

File: functions/go/render-helm-chart/README.md
Comment (mengqiy) [inline functions/go/render-helm-chart/README.md:3]: Can we move all of these links to the bottom of the file?

File: functions/go/render-helm-chart/README.md
Comment (mengqiy) [inline functions/go/render-helm-chart/README.md:103]: Why change this?

File: functions/go/render-helm-chart/README.md
Comment (mengqiy) [inline functions/go/render-helm-chart/README.md:120]: Same here

File: functions/go/render-helm-chart/third_party/sigs.k8s.io/kustomize/api/types/helmchartargs.go
Comment (mengqiy) [inline functions/go/render-helm-chart/third_party/sigs.k8s.io/kustomize/api/types/helmchartargs.go:72]: Why change this?

**Comment by mengqiy** [comment]
The CheckSite CI is failing with the following error:
```
Starting check.
Broken links found.
[sameSite]	❌	http://localhost:3001/master//README.md [x1] {404}
```

@etefera any clue which is file is causing the problem?

File: functions/go/render-helm-chart/README.md
Comment (natasha41575) [reply functions/go/render-helm-chart/README.md:103]: The include crds example was failing, saying that it cannot marshal string to bool

File: examples/render-helm-chart-crds/.expected/config.yaml
Comment (natasha41575) [reply examples/render-helm-chart-crds/.expected/config.yaml:19]: We can't tweak the fields in the input because the input is a remote file. One option I can try is to copy the charts locally and tweak the input.

File: examples/render-helm-chart-crds/.expected/config.yaml
Comment (natasha41575) [reply examples/render-helm-chart-crds/.expected/config.yaml:17]: > Which version of kpt are you using ? kpt version on your local should have version beta2 or above.

Beta3. I'll do some more debugging to see if the problem is something on my end. Thanks for adding the test!

File: functions/go/render-helm-chart/README.md
Comment (natasha41575) [reply functions/go/render-helm-chart/README.md:97]: Changing it back to bool and for it to marshal to map[string]string and copy all the fields

File: examples/render-helm-chart-remote-values-file/README.md
Comment (natasha41575) [reply examples/render-helm-chart-remote-values-file/README.md]: changed!

File: examples/render-helm-chart-kustomize-inline-values/README.md
Comment (natasha41575) [reply examples/render-helm-chart-kustomize-inline-values/README.md]: updated the readme

File: examples/render-helm-chart-crds/.expected/config.yaml
Comment (mengqiy) [inline examples/render-helm-chart-crds/.expected/config.yaml:1]: It seems this file only contains comments.
Should we delete this file given that we have an issue opened for it?

File: examples/render-helm-chart-crds/README.md
Comment (mengqiy) [inline examples/render-helm-chart-crds/README.md:31]: I'm curious why there is a directory here? Is this because the original template has it?

File: examples/render-helm-chart-kustomize-inline-values/.krmignore
Comment (mengqiy) [inline examples/render-helm-chart-kustomize-inline-values/.krmignore:2]: I guess you need to update this.

File: examples/render-helm-chart-kustomize-inline-values/README.md
Comment (mengqiy) [inline examples/render-helm-chart-kustomize-inline-values/README.md:51]: > You should also be able to find `def releaseNamespace = ""` in your output

What's this? I'm not following

File: functions/go/render-helm-chart/main.go
Comment (mengqiy) [inline functions/go/render-helm-chart/main.go:92]: Can you define `ConfigMap` and `RenderHelmChart` as const?

File: examples/render-helm-chart-crds/README.md
Comment (natasha41575) [reply examples/render-helm-chart-crds/README.md:31]: I am actually not sure why this directory is here. I am using the same example as the kustomize test (https://github.com/kubernetes-sigs/kustomize/blob/master/plugin/builtin/helmchartinflationgenerator/HelmChartInflationGenerator_test.go). I think it is using this chart https://github.com/hashicorp/terraform-helm. I am reading through the chart to see if I can understand why it creates this directory.

File: examples/render-helm-chart-kustomize-inline-values/README.md
Comment (natasha41575) [reply examples/render-helm-chart-kustomize-inline-values/README.md:51]: The output contains the following substring:

```
  strategy:
    jenkinsPipelineStrategy:
      jenkinsfile: |-
        def helmName = "helm-v3.1.0-linux-amd64.tar.gz"
        def chartName = "metadata-curator"
        def chartRepo = "http://bcgov.github.io/helm-charts"
        def releaseName  = "mc"
        def releaseNamespace = ""
```

Without the inline values setting `releaseNamespace: ""`, the output would instead have the following: 

```
  strategy:
    jenkinsPipelineStrategy:
      jenkinsfile: |-
        def helmName = "helm-v3.1.0-linux-amd64.tar.gz"
        def chartName = "metadata-curator"
        def chartRepo = "http://bcgov.github.io/helm-charts"
        def releaseName  = "mc"
        def releaseNamespace = "mc"
```

So finding the line `def releaseNamespace = ""` in the output indicates that the correct values are being used.

File: examples/render-helm-chart-crds/.expected/config.yaml
Comment (natasha41575) [reply examples/render-helm-chart-crds/.expected/config.yaml]: deleted

File: examples/render-helm-chart-kustomize-inline-values/.krmignore
Comment (natasha41575) [reply examples/render-helm-chart-kustomize-inline-values/.krmignore]: done

File: functions/go/render-helm-chart/main.go
Comment (natasha41575) [reply functions/go/render-helm-chart/main.go]: done

File: examples/render-helm-chart-kustomize-inline-values/README.md
Comment (natasha41575) [reply examples/render-helm-chart-kustomize-inline-values/README.md]: I updated the wording slightly in the README. Does it help?

File: examples/render-helm-chart-local/.expected/exec.sh
Comment (mengqiy) [inline examples/render-helm-chart-local/.expected/exec.sh:3]: > --image-pull-policy ifNotPresent

Can you please ensure all `render-helm-chart` examples are using `--image-pull-policy never`

File: functions/go/render-helm-chart/main.go
Comment (mengqiy) [inline functions/go/render-helm-chart/main.go:41]: The following is missing, please add.
```
cmd.Example = generated.RenderHelmChartExample
```

**Review by mengqiy — COMMENTED** [review]
LGTM otherwise.

File: examples/render-helm-chart-local/.expected/exec.sh
Comment (natasha41575) [reply examples/render-helm-chart-local/.expected/exec.sh]: sorry about that, checked them all and updated

File: functions/go/render-helm-chart/main.go
Comment (natasha41575) [reply functions/go/render-helm-chart/main.go:41]: done

**Review by mengqiy — APPROVED** [review]

**Pull request merged**
